File: apps/web/src/agent/evalInput.ts
import type { Message, CardInstance, CardSpec } from "@mind-imprint/contracts";
import { SummonCardCall, serializeCardForRefeed } from "@mind-imprint/contracts";

export interface AssembleEvalInputOptions {
  messages: Message[];
  cards: CardInstance[];
  registry: Record<string, CardSpec>;
}

/**
 * Pure function: assembles the evaluation LLM input text from a task's
 * conversation messages and card instances.
 *
 * Returns a plain string with two sections:
 *   ## 对话 — readable conversation transcript
 *   ## 工具卡 — per-card envelope dump (name + status + filled content + event_trace count)
 */
export function assembleEvalInput(opts: AssembleEvalInputOptions): string {
  const { messages, cards, registry } = opts;

  const parts: string[] = [];

  // ── Section 1: Conversation transcript ──────────────────────────────────

  parts.push("## 对话");
  parts.push("");

  for (const m of messages) {
    if (m.role === "user") {
      parts.push(`学生：${m.content}`);
    } else if (m.role === "assistant") {
      const parsed = SummonCardCall.safeParse(m.tool_call);
      if (parsed.success) {
        // Message carries a summon_card proposal
        parts.push(`陪练（建议工具卡）：${m.content}`);
      } else {
        parts.push(`陪练：${m.content}`);
      }
    }
    // system and tool messages are skipped in the eval transcript
  }

  // ── Section 2: Tool card envelopes ──────────────────────────────────────

  parts.push("");
  parts.push("## 工具卡");

  if (cards.length === 0) {
    parts.push("");
    parts.push("（本次对话未触发工具卡）");
  }

  for (const card of cards) {
    const spec = registry[card.card_id];
    const displayName = spec ? spec.name : card.card_id;
    const statusLabel = card.status === "skipped" ? "跳过" : card.status;

    parts.push("");
    parts.push(`【${displayName}】 ${statusLabel}`);

    if (card.status === "completed" || card.status === "skipped") {
      if (spec) {
        const payload = serializeCardForRefeed(spec, card);
        parts.push(JSON.stringify(payload));
      } else {
        // No spec in registry — emit a minimal envelope noting the status
        const fallback = {
          card_id: card.card_id,
          card_name: card.card_id,
          status: card.status,
        };
        if (card.status === "skipped") {
          parts.push(`（跳过，无规格信息）`);
        }
        parts.push(JSON.stringify(fallback));
      }
    }

    parts.push(`（共 ${card.event_trace.length} 个操作事件）`);
  }

  return parts.join("\n");
}

File: apps/api/internal/agent/evalinput.go
package agent

import (
	"encoding/json"
	"strconv"
	"strings"

	"mindimprint/api/internal/cards"
)

// fallbackCard is an ordered struct used when no spec is found in the registry,
// preserving field order card_id, card_name, status in the JSON output.
// (map[string]string does NOT preserve order in Go.)
type fallbackCard struct {
	CardID   string `json:"card_id"`
	CardName string `json:"card_name"`
	Status   string `json:"status"`
}

// AssembleEvalInput ports the TS assembleEvalInput verbatim: a "## 对话"
// transcript (学生/陪练 lines, 陪练（建议工具卡） when a summon_card rode along)
// followed by a "## 工具卡" dump (per card: display name + status label, the
// serialized refeed payload for completed/skipped, and the event-count line).
func AssembleEvalInput(messages []StoredMessage, cardInsts []CardInstance, specByID func(string) (cards.Spec, bool)) string {
	var parts []string

	// ── Section 1: Conversation transcript ──────────────────────────────────

	parts = append(parts, "## 对话")
	parts = append(parts, "")

	for _, m := range messages {
		switch m.Role {
		case "user":
			parts = append(parts, "学生："+m.Content)
		case "assistant":
			if m.ToolCall != nil && m.ToolCall.Name == "summon_card" {
				parts = append(parts, "陪练（建议工具卡）："+m.Content)
			} else {
				parts = append(parts, "陪练："+m.Content)
			}
		}
		// system and tool messages are skipped in the eval transcript
	}

	// ── Section 2: Tool card envelopes ──────────────────────────────────────

	parts = append(parts, "")
	parts = append(parts, "## 工具卡")

	if len(cardInsts) == 0 {
		parts = append(parts, "")
		parts = append(parts, "（本次对话未触发工具卡）")
	}

	for _, c := range cardInsts {
		spec, hasSpec := specByID(c.CardID)
		displayName := c.CardID
		if hasSpec {
			displayName = spec.Name
		}
		statusLabel := c.Status
		if c.Status == "skipped" {
			statusLabel = "跳过"
		}

		parts = append(parts, "")
		parts = append(parts, "【"+displayName+"】 "+statusLabel)

		if c.Status == "completed" || c.Status == "skipped" {
			if hasSpec {
				b, _ := json.Marshal(SerializeCardForRefeed(spec, c))
				parts = append(parts, string(b))
			} else {
				// No spec in registry — emit a minimal envelope noting the status
				if c.Status == "skipped" {
					parts = append(parts, "（跳过，无规格信息）")
				}
				b, _ := json.Marshal(fallbackCard{CardID: c.CardID, CardName: c.CardID, Status: c.Status})
				parts = append(parts, string(b))
			}
		}

		parts = append(parts, "（共 "+strconv.Itoa(c.EventTraceLen)+" 个操作事件）")
	}

	return strings.Join(parts, "\n")
}
